File: lib/go/protochain/solana/program/token/v1/metadata_test.go
package token_v1

import (
	"encoding/json"
	"testing"
)

// ---------------------------------------------------------------------------
// Fixtures – taken verbatim from the Metaplex Token Standard documentation
// ---------------------------------------------------------------------------

const fungibleUSDC = `{
  "name": "USD Coin",
  "symbol": "USDC",
  "description": "Fully reserved fiat-backed stablecoin created by Circle.",
  "image": "https://www.circle.com/hs-fs/hubfs/sundaes/USDC.png?width=540&height=540&name=USDC.png"
}`

const fungibleAssetSword = `{
  "name": "SolanaGame Steel Sword",
  "symbol": "SG-SS-1",
  "description": "SolanaGame steel sword available after Level 4",
  "image": "https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg",
  "animation_url": "https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb",
  "external_url": "https://SolanaGame.io",
  "attributes": [
    { "trait_type": "attack", "value": "4" },
    { "trait_type": "defense", "value": "3" },
    { "trait_type": "durability", "value": "47" },
    { "trait_type": "components", "value": "iron: 10; carbon: 1; wood: 2" }
  ]
}`

const nftWithDeprecatedFields = `{
  "name": "SolanaArtProject #1",
  "description": "Generative art on Solana.",
  "image": "https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg",
  "animation_url": "https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb",
  "external_url": "https://example.com",
  "attributes": [
    { "trait_type": "trait1", "value": "value1" },
    { "trait_type": "trait2", "value": "value2" }
  ],
  "properties": {
    "files": [
      { "uri": "https://www.arweave.net/abcd5678?ext=png", "type": "image/png" },
      { "uri": "https://watch.videodelivery.net/9876jkl", "type": "unknown", "cdn": true },
      { "uri": "https://www.arweave.net/efgh1234?ext=mp4", "type": "video/mp4" }
    ],
    "category": "video",
    "creators": [
      { "address": "xEtQ9Fpv62qdc1GYfpNReMasVTe9YW5bHJwfVKqo72u", "share": 100 }
    ]
  }
}`

const nftClean = `{
  "name": "SolanaArtProject #1",
  "description": "Generative art on Solana.",
  "image": "https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg",
  "animation_url": "https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb",
  "external_url": "https://example.com",
  "attributes": [
    { "trait_type": "trait1", "value": "value1" },
    { "trait_type": "trait2", "value": "value2" }
  ],
  "properties": {
    "files": [
      { "uri": "https://www.arweave.net/abcd5678?ext=png", "type": "image/png" },
      { "uri": "https://watch.videodelivery.net/9876jkl", "type": "unknown", "cdn": true },
      { "uri": "https://www.arweave.net/efgh1234?ext=mp4", "type": "video/mp4" }
    ],
    "category": "video"
  }
}`

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func mustParse(t *testing.T, data string) *UniversalTokenMetadata {
	t.Helper()
	m, err := ParseUniversalTokenMetadata([]byte(data))
	if err != nil {
		t.Fatalf("ParseUniversalTokenMetadata failed: %v", err)
	}
	return m
}

func ptrBool(v bool) *bool          { return &v }
func ptrFloat64(v float64) *float64 { return &v }
func ptrInt(v int) *int             { return &v }

// ---------------------------------------------------------------------------
// Tests – Fungible (USDC)
// ---------------------------------------------------------------------------

func TestParseFungibleUSDC(t *testing.T) {
	m := mustParse(t, fungibleUSDC)

	assertEqual(t, "name", m.Name, "USD Coin")
	assertEqual(t, "symbol", m.Symbol, "USDC")
	assertEqual(t, "description", m.Description, "Fully reserved fiat-backed stablecoin created by Circle.")
	assertEqual(t, "image", m.Image, "https://www.circle.com/hs-fs/hubfs/sundaes/USDC.png?width=540&height=540&name=USDC.png")
}

func TestParseFungibleUSDC_OptionalFieldsAbsent(t *testing.T) {
	m := mustParse(t, fungibleUSDC)

	assertEqual(t, "animation_url", m.AnimationURL, "")
	assertEqual(t, "external_url", m.ExternalURL, "")
	assertEqual(t, "background_color", m.BackgroundColor, "")
	assertNil(t, "attributes", m.Attributes)
	assertNil(t, "properties", m.Properties)
	assertNil(t, "collection", m.Collection)
	assertNil(t, "seller_fee_basis_points", m.SellerFeeBasisPoints)
}

// ---------------------------------------------------------------------------
// Tests – FungibleAsset (Steel Sword)
// ---------------------------------------------------------------------------

func TestParseFungibleAssetSword_TopLevelFields(t *testing.T) {
	m := mustParse(t, fungibleAssetSword)

	assertEqual(t, "name", m.Name, "SolanaGame Steel Sword")
	assertEqual(t, "symbol", m.Symbol, "SG-SS-1")
	assertEqual(t, "description", m.Description, "SolanaGame steel sword available after Level 4")
	assertEqual(t, "image", m.Image, "https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg")
	assertEqual(t, "animation_url", m.AnimationURL, "https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb")
	assertEqual(t, "external_url", m.ExternalURL, "https://SolanaGame.io")
}

func TestParseFungibleAssetSword_Attributes(t *testing.T) {
	m := mustParse(t, fungibleAssetSword)

	if len(m.Attributes) != 4 {
		t.Fatalf("expected 4 attributes, got %d", len(m.Attributes))
	}

	assertAttributeString(t, m.Attributes[0], "attack", "4")
	assertAttributeString(t, m.Attributes[3], "components", "iron: 10; carbon: 1; wood: 2")
}

func TestParseFungibleAssetSword_NoProperties(t *testing.T) {
	m := mustParse(t, fungibleAssetSword)

	assertNil(t, "properties", m.Properties)
	assertNil(t, "collection", m.Collection)
}

// ---------------------------------------------------------------------------
// Tests – NonFungible with deprecated fields
// ---------------------------------------------------------------------------

func TestParseNFTDeprecated_Files(t *testing.T) {
	m := mustParse(t, nftWithDeprecatedFields)

	if m.Properties == nil {
		t.Fatal("expected properties to be present")
	}
	if len(m.Properties.Files) != 3 {
		t.Fatalf("expected 3 files, got %d", len(m.Properties.Files))
	}

	f0 := m.Properties.Files[0]
	assertEqual(t, "files[0].uri", f0.URI, "https://www.arweave.net/abcd5678?ext=png")
	assertEqual(t, "files[0].type", f0.Type, "image/png")
	assertNil(t, "files[0].cdn", f0.CDN)

	f1 := m.Properties.Files[1]
	assertEqual(t, "files[1].uri", f1.URI, "https://watch.videodelivery.net/9876jkl")
	assertEqual(t, "files[1].type", f1.Type, "unknown")
	assertDeepEqual(t, "files[1].cdn", f1.CDN, ptrBool(true))
}

func TestParseNFTDeprecated_Category(t *testing.T) {
	m := mustParse(t, nftWithDeprecatedFields)

	assertEqual(t, "category", string(m.Properties.Category), "video")
}

func TestParseNFTDeprecated_Creators(t *testing.T) {
	m := mustParse(t, nftWithDeprecatedFields)

	if len(m.Properties.Creators) != 1 {
		t.Fatalf("expected 1 creator, got %d", len(m.Properties.Creators))
	}
	assertEqual(t, "creators[0].address", m.Properties.Creators[0].Address, "xEtQ9Fpv62qdc1GYfpNReMasVTe9YW5bHJwfVKqo72u")
	assertEqual(t, "creators[0].share", m.Properties.Creators[0].Share, 100)
}

func TestParseNFTDeprecated_NoTopLevelCollection(t *testing.T) {
	m := mustParse(t, nftWithDeprecatedFields)

	assertNil(t, "collection", m.Collection)
}

// ---------------------------------------------------------------------------
// Tests – NonFungible (clean)
// ---------------------------------------------------------------------------

func TestParseNFTClean_AllFields(t *testing.T) {
	m := mustParse(t, nftClean)

	assertEqual(t, "name", m.Name, "SolanaArtProject #1")
	assertEqual(t, "description", m.Description, "Generative art on Solana.")
	assertEqual(t, "animation_url", m.AnimationURL, "https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb")
	assertEqual(t, "external_url", m.ExternalURL, "https://example.com")
}

func TestParseNFTClean_PropertiesWithoutCreators(t *testing.T) {
	m := mustParse(t, nftClean)

	if m.Properties == nil {
		t.Fatal("expected properties to be present")
	}
	if len(m.Properties.Files) != 3 {
		t.Fatalf("expected 3 files, got %d", len(m.Properties.Files))
	}
	assertEqual(t, "category", string(m.Properties.Category), "video")
	assertNil(t, "creators", m.Properties.Creators)
}

func TestParseNFTClean_SymbolAbsent(t *testing.T) {
	m := mustParse(t, nftClean)

	assertEqual(t, "symbol", m.Symbol, "")
}

// ---------------------------------------------------------------------------
// Tests – Edge cases
// ---------------------------------------------------------------------------

func TestParseInvalidJSON(t *testing.T) {
	_, err := ParseUniversalTokenMetadata([]byte(`not json`))
	if err == nil {
		t.Fatal("expected error for invalid JSON")
	}
}

func TestParseEmptyObject(t *testing.T) {
	m := mustParse(t, `{}`)

	assertEqual(t, "name", m.Name, "")
	assertEqual(t, "description", m.Description, "")
	assertEqual(t, "image", m.Image, "")
}

func TestParseNumericAttributeValue(t *testing.T) {
	m := mustParse(t, `{
		"name": "test", "description": "test", "image": "test",
		"attributes": [{"trait_type": "Level", "value": 5}]
	}`)

	f, ok := m.Attributes[0].Value.Float64Value()
	if !ok {
		t.Fatal("expected float64 value")
	}
	assertEqualFloat(t, "value", f, 5.0)
}

func TestParseBooleanAttributeValue(t *testing.T) {
	m := mustParse(t, `{
		"name": "test", "description": "test", "image": "test",
		"attributes": [{"trait_type": "Legendary", "value": true}]
	}`)

	b, ok := m.Attributes[0].Value.BoolValue()
	if !ok {
		t.Fatal("expected bool value")
	}
	if !b {
		t.Fatal("expected true")
	}
}

func TestParseDisplayTypeAndMaxValue(t *testing.T) {
	m := mustParse(t, `{
		"name": "test", "description": "test", "image": "test",
		"attributes": [
			{"trait_type": "Level", "value": 5, "display_type": "number", "max_value": 10},
			{"trait_type": "Power Boost", "value": 40, "display_type": "boost_percentage"},
			{"trait_type": "Birthday", "value": 1546360800, "display_type": "date"}
		]
	}`)

	if len(m.Attributes) != 3 {
		t.Fatalf("expected 3 attributes, got %d", len(m.Attributes))
	}

	a0 := m.Attributes[0]
	assertEqual(t, "a0.display_type", string(a0.DisplayType), "number")
	assertDeepEqual(t, "a0.max_value", a0.MaxValue, ptrFloat64(10))

	a1 := m.Attributes[1]
	assertEqual(t, "a1.display_type", string(a1.DisplayType), "boost_percentage")
	assertNil(t, "a1.max_value", a1.MaxValue)

	a2 := m.Attributes[2]
	assertEqual(t, "a2.display_type", string(a2.DisplayType), "date")
}

func TestParseTopLevelCollection(t *testing.T) {
	m := mustParse(t, `{
		"name": "test", "description": "test", "image": "test",
		"collection": {"name": "My Collection", "family": "My Family"}
	}`)

	if m.Collection == nil {
		t.Fatal("expected collection to be present")
	}
	assertEqual(t, "collection.name", m.Collection.Name, "My Collection")
	assertEqual(t, "collection.family", m.Collection.Family, "My Family")
}

func TestParseSellerFeeBasisPoints(t *testing.T) {
	m := mustParse(t, `{
		"name": "test", "description": "test", "image": "test",
		"seller_fee_basis_points": 500
	}`)

	assertDeepEqual(t, "seller_fee_basis_points", m.SellerFeeBasisPoints, ptrInt(500))
}

func TestParseBackgroundColor(t *testing.T) {
	m := mustParse(t, `{
		"name": "test", "description": "test", "image": "test",
		"background_color": "FF0000"
	}`)

	assertEqual(t, "background_color", m.BackgroundColor, "FF0000")
}

func TestAttributeValueRoundTrip(t *testing.T) {
	m := mustParse(t, `{
		"name": "test", "description": "test", "image": "test",
		"attributes": [
			{"trait_type": "s", "value": "hello"},
			{"trait_type": "n", "value": 42},
			{"trait_type": "b", "value": false}
		]
	}`)

	data, err := json.Marshal(m)
	if err != nil {
		t.Fatalf("marshal failed: %v", err)
	}

	m2 := mustParse(t, string(data))
	if len(m2.Attributes) != 3 {
		t.Fatalf("expected 3 attributes after round-trip, got %d", len(m2.Attributes))
	}

	s, ok := m2.Attributes[0].Value.StringValue()
	if !ok || s != "hello" {
		t.Fatalf("expected string 'hello', got %v", m2.Attributes[0].Value.Raw())
	}

	f, ok := m2.Attributes[1].Value.Float64Value()
	if !ok || f != 42.0 {
		t.Fatalf("expected float64 42, got %v", m2.Attributes[1].Value.Raw())
	}

	b, ok := m2.Attributes[2].Value.BoolValue()
	if !ok || b {
		t.Fatalf("expected bool false, got %v", m2.Attributes[2].Value.Raw())
	}
}

func TestAttributeValueString(t *testing.T) {
	v := AttributeValue{raw: "hello"}
	assertEqual(t, "String()", v.String(), "hello")

	v2 := AttributeValue{raw: 42.0}
	assertEqual(t, "String()", v2.String(), "42")
}

// ---------------------------------------------------------------------------
// Test helpers
// ---------------------------------------------------------------------------

func assertEqual[T comparable](t *testing.T, field string, got, want T) {
	t.Helper()
	if got != want {
		t.Errorf("%s: got %v, want %v", field, got, want)
	}
}

func assertEqualFloat(t *testing.T, field string, got, want float64) {
	t.Helper()
	if got != want {
		t.Errorf("%s: got %v, want %v", field, got, want)
	}
}

func assertNil[T any](t *testing.T, field string, v T) {
	t.Helper()
	if !isNil(v) {
		t.Errorf("%s: expected nil, got %v", field, v)
	}
}

func assertDeepEqual[T any](t *testing.T, field string, got, want T) {
	t.Helper()
	gotJSON, _ := json.Marshal(got)
	wantJSON, _ := json.Marshal(want)
	if string(gotJSON) != string(wantJSON) {
		t.Errorf("%s: got %s, want %s", field, gotJSON, wantJSON)
	}
}

func assertAttributeString(t *testing.T, attr MetadataAttribute, wantTrait, wantValue string) {
	t.Helper()
	assertEqual(t, "trait_type", attr.TraitType, wantTrait)
	s, ok := attr.Value.StringValue()
	if !ok {
		t.Errorf("expected string value for trait %q, got %T", wantTrait, attr.Value.Raw())
		return
	}
	assertEqual(t, "value", s, wantValue)
}

// isNil checks if an interface value is nil, handling typed nils.
func isNil(v any) bool {
	if v == nil {
		return true
	}
	// Use json.Marshal trick: typed nil pointers/slices marshal to "null"
	data, _ := json.Marshal(v)
	return string(data) == "null"
}

File: lib/ts-web/src/protochain/solana/program/token/v1/metadata.test.ts
import { describe, expect, it } from 'vitest';

import { MetadataParser } from './metadata';

// ---------------------------------------------------------------------------
// Fixtures – taken verbatim from the Metaplex Token Standard documentation
// ---------------------------------------------------------------------------

const FUNGIBLE_USDC = {
  name: 'USD Coin',
  symbol: 'USDC',
  description: 'Fully reserved fiat-backed stablecoin created by Circle.',
  image: 'https://www.circle.com/hs-fs/hubfs/sundaes/USDC.png?width=540&height=540&name=USDC.png',
};

const FUNGIBLE_ASSET_SWORD = {
  name: 'SolanaGame Steel Sword',
  symbol: 'SG-SS-1',
  description: 'SolanaGame steel sword available after Level 4',
  image: 'https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg',
  animation_url: 'https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb',
  external_url: 'https://SolanaGame.io',
  attributes: [
    { trait_type: 'attack', value: '4' },
    { trait_type: 'defense', value: '3' },
    { trait_type: 'durability', value: '47' },
    { trait_type: 'components', value: 'iron: 10; carbon: 1; wood: 2' },
  ],
};

const NFT_WITH_DEPRECATED_FIELDS = {
  name: 'SolanaArtProject #1',
  description: 'Generative art on Solana.',
  image: 'https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg',
  animation_url: 'https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb',
  external_url: 'https://example.com',
  attributes: [
    { trait_type: 'trait1', value: 'value1' },
    { trait_type: 'trait2', value: 'value2' },
  ],
  properties: {
    files: [
      { uri: 'https://www.arweave.net/abcd5678?ext=png', type: 'image/png' },
      { uri: 'https://watch.videodelivery.net/9876jkl', type: 'unknown', cdn: true },
      { uri: 'https://www.arweave.net/efgh1234?ext=mp4', type: 'video/mp4' },
    ],
    category: 'video',
    collection: { name: 'Solflare X NFT', family: 'Solflare' },
    creators: [{ address: 'xEtQ9Fpv62qdc1GYfpNReMasVTe9YW5bHJwfVKqo72u', share: 100 }],
  },
};

const NFT_CLEAN = {
  name: 'SolanaArtProject #1',
  description: 'Generative art on Solana.',
  image: 'https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg',
  animation_url: 'https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb',
  external_url: 'https://example.com',
  attributes: [
    { trait_type: 'trait1', value: 'value1' },
    { trait_type: 'trait2', value: 'value2' },
  ],
  properties: {
    files: [
      { uri: 'https://www.arweave.net/abcd5678?ext=png', type: 'image/png' },
      { uri: 'https://watch.videodelivery.net/9876jkl', type: 'unknown', cdn: true },
      { uri: 'https://www.arweave.net/efgh1234?ext=mp4', type: 'video/mp4' },
    ],
    category: 'video',
  },
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('MetadataParser.fromJSON', () => {
  // ---- Fungible (minimal) ------------------------------------------------

  describe('Fungible token (USDC)', () => {
    it('parses required fields', () => {
      const meta = MetadataParser.fromJSON(FUNGIBLE_USDC);

      expect(meta.name).toBe('USD Coin');
      expect(meta.symbol).toBe('USDC');
      expect(meta.description).toBe('Fully reserved fiat-backed stablecoin created by Circle.');
      expect(meta.image).toBe(
        'https://www.circle.com/hs-fs/hubfs/sundaes/USDC.png?width=540&height=540&name=USDC.png',
      );
    });

    it('leaves optional fields undefined', () => {
      const meta = MetadataParser.fromJSON(FUNGIBLE_USDC);

      expect(meta.animation_url).toBeUndefined();
      expect(meta.external_url).toBeUndefined();
      expect(meta.background_color).toBeUndefined();
      expect(meta.attributes).toBeUndefined();
      expect(meta.properties).toBeUndefined();
      expect(meta.collection).toBeUndefined();
      expect(meta.seller_fee_basis_points).toBeUndefined();
    });
  });

  // ---- FungibleAsset (with attributes) -----------------------------------

  describe('FungibleAsset (Steel Sword)', () => {
    it('parses all top-level fields', () => {
      const meta = MetadataParser.fromJSON(FUNGIBLE_ASSET_SWORD);

      expect(meta.name).toBe('SolanaGame Steel Sword');
      expect(meta.symbol).toBe('SG-SS-1');
      expect(meta.description).toBe('SolanaGame steel sword available after Level 4');
      expect(meta.image).toBe(
        'https://arweave.net/26YdhY_eAzv26YdhY1uu9uiA3nmDZYwP8MwZAultcE?ext=jpeg',
      );
      expect(meta.animation_url).toBe(
        'https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb',
      );
      expect(meta.external_url).toBe('https://SolanaGame.io');
    });

    it('parses string-valued attributes', () => {
      const meta = MetadataParser.fromJSON(FUNGIBLE_ASSET_SWORD);

      expect(meta.attributes).toHaveLength(4);
      expect(meta.attributes![0]).toEqual({ trait_type: 'attack', value: '4' });
      expect(meta.attributes![3]).toEqual({
        trait_type: 'components',
        value: 'iron: 10; carbon: 1; wood: 2',
      });
    });

    it('leaves properties and collection undefined when absent', () => {
      const meta = MetadataParser.fromJSON(FUNGIBLE_ASSET_SWORD);

      expect(meta.properties).toBeUndefined();
      expect(meta.collection).toBeUndefined();
    });
  });

  // ---- NonFungible with deprecated fields --------------------------------

  describe('NonFungible with deprecated collection & creators', () => {
    it('parses properties.files with cdn flag', () => {
      const meta = MetadataParser.fromJSON(NFT_WITH_DEPRECATED_FIELDS);

      expect(meta.properties).toBeDefined();
      expect(meta.properties!.files).toHaveLength(3);
      expect(meta.properties!.files![0]).toEqual({
        uri: 'https://www.arweave.net/abcd5678?ext=png',
        type: 'image/png',
      });
      expect(meta.properties!.files![1]).toEqual({
        uri: 'https://watch.videodelivery.net/9876jkl',
        type: 'unknown',
        cdn: true,
      });
    });

    it('parses properties.category', () => {
      const meta = MetadataParser.fromJSON(NFT_WITH_DEPRECATED_FIELDS);

      expect(meta.properties!.category).toBe('video');
    });

    it('parses deprecated properties.creators', () => {
      const meta = MetadataParser.fromJSON(NFT_WITH_DEPRECATED_FIELDS);

      expect(meta.properties!.creators).toHaveLength(1);
      expect(meta.properties!.creators![0]).toEqual({
        address: 'xEtQ9Fpv62qdc1GYfpNReMasVTe9YW5bHJwfVKqo72u',
        share: 100,
      });
    });

    it('ignores unknown nested keys (collection inside properties)', () => {
      // The `collection` key sits inside `properties` in the fixture – it is
      // not a recognised field of MetadataProperties, so the parser should
      // silently drop it. The top-level `collection` should be undefined.
      const meta = MetadataParser.fromJSON(NFT_WITH_DEPRECATED_FIELDS);

      expect(meta.collection).toBeUndefined();
    });
  });

  // ---- NonFungible without deprecated fields -----------------------------

  describe('NonFungible (clean, no deprecated fields)', () => {
    it('parses all fields correctly', () => {
      const meta = MetadataParser.fromJSON(NFT_CLEAN);

      expect(meta.name).toBe('SolanaArtProject #1');
      expect(meta.description).toBe('Generative art on Solana.');
      expect(meta.animation_url).toBe(
        'https://arweave.net/ZAultcE_eAzv26YdhY1uu9uiA3nmDZYwP8MwuiA3nm?ext=glb',
      );
      expect(meta.external_url).toBe('https://example.com');
    });

    it('parses properties without creators', () => {
      const meta = MetadataParser.fromJSON(NFT_CLEAN);

      expect(meta.properties).toBeDefined();
      expect(meta.properties!.files).toHaveLength(3);
      expect(meta.properties!.category).toBe('video');
      expect(meta.properties!.creators).toBeUndefined();
    });

    it('symbol defaults to undefined when absent', () => {
      const meta = MetadataParser.fromJSON(NFT_CLEAN);

      expect(meta.symbol).toBeUndefined();
    });
  });

  // ---- Edge cases & defensive parsing ------------------------------------

  describe('edge cases', () => {
    it('throws on null input', () => {
      expect(() => MetadataParser.fromJSON(null)).toThrow('input must be a non-null object');
    });

    it('throws on non-object input', () => {
      expect(() => MetadataParser.fromJSON('not an object')).toThrow(
        'input must be a non-null object',
      );
    });

    it('coerces missing required fields to empty strings', () => {
      const meta = MetadataParser.fromJSON({});

      expect(meta.name).toBe('');
      expect(meta.description).toBe('');
      expect(meta.image).toBe('');
    });

    it('handles numeric attribute values', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        attributes: [{ trait_type: 'Level', value: 5 }],
      });

      expect(meta.attributes![0].value).toBe(5);
      expect(typeof meta.attributes![0].value).toBe('number');
    });

    it('handles boolean attribute values', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        attributes: [{ trait_type: 'Legendary', value: true }],
      });

      expect(meta.attributes![0].value).toBe(true);
      expect(typeof meta.attributes![0].value).toBe('boolean');
    });

    it('parses display_type and max_value on attributes', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        attributes: [
          { trait_type: 'Level', value: 5, display_type: 'number', max_value: 10 },
          { trait_type: 'Power Boost', value: 40, display_type: 'boost_percentage' },
          { trait_type: 'Birthday', value: 1546360800, display_type: 'date' },
        ],
      });

      expect(meta.attributes![0]).toEqual({
        trait_type: 'Level',
        value: 5,
        display_type: 'number',
        max_value: 10,
      });
      expect(meta.attributes![1]).toEqual({
        trait_type: 'Power Boost',
        value: 40,
        display_type: 'boost_percentage',
      });
      expect(meta.attributes![2]).toEqual({
        trait_type: 'Birthday',
        value: 1546360800,
        display_type: 'date',
      });
    });

    it('parses top-level collection (name + family)', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        collection: { name: 'My Collection', family: 'My Family' },
      });

      expect(meta.collection).toEqual({ name: 'My Collection', family: 'My Family' });
    });

    it('parses seller_fee_basis_points', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        seller_fee_basis_points: 500,
      });

      expect(meta.seller_fee_basis_points).toBe(500);
    });

    it('parses background_color', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        background_color: 'FF0000',
      });

      expect(meta.background_color).toBe('FF0000');
    });

    it('skips malformed entries in attributes array', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        attributes: [null, 'invalid', { trait_type: 'valid', value: 'yes' }],
      });

      expect(meta.attributes).toHaveLength(1);
      expect(meta.attributes![0].trait_type).toBe('valid');
    });

    it('skips malformed entries in properties.files array', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        properties: {
          files: [null, 42, { uri: 'https://example.com/img.png', type: 'image/png' }],
        },
      });

      expect(meta.properties!.files).toHaveLength(1);
      expect(meta.properties!.files![0].uri).toBe('https://example.com/img.png');
    });

    it('ignores non-finite seller_fee_basis_points', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        seller_fee_basis_points: 'not a number',
      });

      expect(meta.seller_fee_basis_points).toBeUndefined();
    });

    it('returns undefined collection when object has neither name nor family', () => {
      const meta = MetadataParser.fromJSON({
        name: 'test',
        description: 'test',
        image: 'test',
        collection: {},
      });

      expect(meta.collection).toBeUndefined();
    });
  });
});
